# 3. gyakorlat

## Dualitás, érzékenységvizsgálat

## Célok:
* Duális LP 'levezetése' és megoldás leolvasása a primál LP megoldásából
* Árnyékárak értelmezése
* Célfüggvényegyütthatók optimalitási tartományának meghatározása
* Mátrixjátékok és dualitás kapcsolata

*1. feladat*

Egy varrónő ingeket és nyakkendőket készít 2 fajta textílből (A és B minőségű), valamint maradékanyagból. Az A minőségű anyagból 24, a B-ből 42 méternyi áll rendelkezésére. Maradékanyagból 4 méternyi készlete van. Egy ing elkészítéséhez 1 méternyi A minőségű és 4 méternyi B minőségű anyagra van szüksége, keletkezik viszont 1 méternyi maradékanyaga. Egy tucat nyakkendő elkészítéséhez 3 méternyi A minőségű és 3 méternyi B minőségű anyagra van szüksége, valamint 1 méternyi maradékanyagra. Egy inget 20000 Forintért, egy tucat nyakkendőt pedig 30000 Forintért tud értékesíteni.
<br>a) Írjunk fel egy lineáris programozási feladatot a varrónő bevételének maximalizására, és oldjuk meg pulp segítségével!

*Megoldás:*
<br> $x_1$ - a legyártott ingek száma
<br> $x_2$ - a legyártott tucat nyakkendők száma
$$\color{blue}{
\begin{array}{rccc}
\max & \phantom{-}20x_1 + 30x_2 & & \\
\text{f.h.} & \phantom{-20}x_1 + \phantom{0}3x_2 & \leq & 24 \\
& \phantom{-0}4x_1 + \phantom{0}3x_2 & \leq & 42 \\
& \phantom{2}-x_1 + \phantom{3}x_2 & \leq & 4 \\
& \phantom{-20}x_1, \phantom{30}x_2 & \geq & 0
\end{array}
}$$

In [ ]:
#pip install pulp # ha még egyszer sem volt telepítve

from pulp import *

varrono = LpProblem('Varrónő', LpMaximize)

x1 = LpVariable('ing', lowBound=0 , cat = 'Continuous')
x2 = LpVariable('nyakkendo', lowBound=0 , cat = 'Continuous')

varrono += 20*x1 + 30*x2

varrono += x1 + 3*x2 <= 24, "A anyag"
varrono += 4*x1 + 3*x2 <= 42, "B anyag"
varrono += - x1 + x2 <= 4, "maradék anyag"

print(varrono)

megoldas = varrono.solve()

if LpStatus[megoldas] == 'Optimal':
    print('Optimális megoldást találtunk.')
    print('Célfüggvény érték: z* = ', value(varrono.objective))
    print('Megoldás: ing* = ', value(x1), ', nyakkendo* = ', value(x2))
elif megoldas == -1:
    print('Az LP-nek nincs megengedett megoldása.')
elif LpStatus[megoldas] == 'Unbounded':
    print('A célfüggvény nem korlátos.')
else:
    print('Az LP megoldása valamely problémába ütközött.')

b) Adjunk meg egy **felső korlátot** a célfüggvény értékére az *első korlát alapján*!

*Megoldás:*
<br>Az $x_1$ együtthatója a célfüggvényben 20, míg a korlátban 1; ugyanez az $x_2$ esetében 30 és 3. Szorozzuk meg a korlátot 20-szal annak érdekében, hogy szét tudjuk bontani az első korlátot magára a célfüggvényre + egy nem-negatív értékre:

$$
20x_1 + 60x_2 = \color{blue}{20x_1 + 30x_2} + 30x_2 \leq 480
$$

Az optimális megoldásban felvett célfüggvény érték az első korlát miatt nem lehet nagyobb, mint 480.

c) Adjunk meg jobb felső korlátot a második és harmadik korlát használatával!

*Megoldás:*
<br>Szorozzuk meg a második korlátot valamilyen $y_2$, míg a harmadikat valamilyen $y_3$ nem-negatív értékkel, majd összegezzük az így kapott korlátokat. Az eredmény:

$$
(4y_2 - y_3)x_1 + (3y_2 + y_3)x_2 \leq 42y_2 + 4y_3
$$

*Megjegyzés: bajban lennénk, ha ez korlátként megjelenne az $x$ és $y$ változókban, hiszen nemlineáris!*
<br>Ezúttal azonban csak egy felső korlát értékének megtalálásához használjuk.

Ahhoz, hogy szétbonthassuk a korábbi módon, $x_1$ együtthatójának (ami $4y_2 - y_3$) legalább 20-nak ($x_1$ célfgv. együtthatója), $x_2$-nél pedig a $3y_2 + y_3$ együtthatónak kell legalább 30-nak lennie. Mindezt úgy, hogy $y_2, y_3 \geq 0$ kell az összegzéshez.

Például $y_2 = 8, y_3 = 6$ esetén

$$26x_1 + 30x_2 = \color{blue}{20x_1 + 30x_2} + 6x_1 \leq 360$$

d) Írjuk fel a legjobb korlátot meghatározó duális LP-t! Adjuk meg egy megoldását!

*Megoldás:*
<br>Innen adódik a **duális** feladat, mely egy LP segítségével megtalálja a legjobb korlátot biztosító $y$ változókat.

$$\color{blue}{
\begin{array}{rcccc}
\max & \phantom{-}20x_1 + 30x_2 & & & \hspace{10mm} \color{red}{\text{(duál változók)}}\\
\text{f.h.} & \phantom{-20}x_1 + \phantom{0}3x_2 & \leq & 24 & \hspace{10mm} \color{red}{(y_1)}\\
& \phantom{-0}4x_1 + \phantom{0}3x_2 & \leq & 42 & \hspace{10mm} \color{red}{(y_2)}\\
& \phantom{2}-x_1 + \phantom{3}x_2 & \leq & 4 & \hspace{10mm} \color{red}{(y_3)}\\
& \phantom{-20}x_1, \phantom{30}x_2 & \geq & 0
\end{array}
}$$

$$\color{red}{
\begin{array}{rcccc}
\min & 24y_1 + 42y_2 + 4y_3 & & & \hspace{10mm} \color{blue}{\text{(primál változók)}}\\
\text{f.h.} & \phantom{24}y_1 + \phantom{2}4y_2 - \phantom{4}y_3 & \geq & 20 & \hspace{10mm} \color{blue}{(x_1)}\\
& \phantom{2}3y_1 + \phantom{4}3y_2 + \phantom{4}y_3 & \geq & 30 & \hspace{10mm} \color{blue}{(x_2)}\\
& \phantom{24}y_1, \phantom{42}y_2, \phantom{4}y_3 & \geq & 0
\end{array}
}$$

* Primál vátozók száma = duál korlátok száma és fordítva
* primál célfüggvény együtthatók = duál jobb oldali konstansok és fordítva
* a primál együtthatómátrix a duál együtthatómátrix transzponáltja

A duál feladat értelmezése: Tegyük fel, hogy egy vállalkozó fel szeretné vásárolni a varrónő alapanyag készletét. Az anyagok árát jelöljék a következő változók
* $y_1$ - az A textil méterének ára
* $y_2$ - a B textil méterének ára
* $y_3$ - a maradékanyag méterének ára

Ekkor a vállalkozó a következő célfüggvényt minimalizálja: $24y_1 + 42y_2 + 4y_3$
* A vállalkozónak az A és B anyag 3-3 méteréért, valamint 1 méternyi maradékanyagért mindenképp többet kell adnia mint 30 ezer Ft
* Ha ennél kevesebbet adna, akkor a varrónő inkább nyakkendőt készítene az anyagból és azt adná el
* Hasonlóan 1 méternyi A textilért és 4 méternyi B textilért többet kell adnia mint 20 ezer Ft plusz 1 méternyi maradékanyag ára

A duál feladat $y^*$ optimális megoldásai a primál korlátokhoz optimális megoldásban tartozó ún. **árnyékárak**. Ezek értékét meg tudjuk határozni a pulp segítségével, amikor megoldjuk az eredeti feladatot (nincs szükség újabb LP megoldásra).

A definiált LpProblem változónk *constraints* függvénye listázza (egy ún. dictionaryben tárolva) a korlátozó feltételeket, a *constraints.items()* függvény segítségével pedig tudunk iterálni is a listaelemeken.

In [ ]:
for name, c in varrono.constraints.items():
    print('y_' + name + ' =', c.pi)

e) Értelmezzük az egyes árnyékárakat! A rendelkezésünkre álló erőforrások közül melyik mennyire értékes? Határozzuk meg melyek az aktív és inaktív feltételek!

*Megoldás:*
<br>Egy méternyi plusz A anyag a varrónő számára $y_1 = 6.6666667$ többletbevételt eredményezne.
<br>Ugyanez a B anyag esetében $y_2 = 3.3333333$.
<br>A maradékanyagból újabb méter nem okoz bevételnövekedést a varrónő számára.
<br> Az első két feltétel aktív, a harmadik feltétel inaktív.
<br> *Megjegyzés: inaktív feltétel árnyékára mindenképpen 0*

f) Az inaktív feltételek esetén: a jobb oldali konstans milyen intervallumban változhat ahhoz, hogy az optimális megoldás ne változzon? Hogyan változik eközben az optimális célfüggvény érték?

*Megoldás:*
<br>$\leq$ korlát lévén a jobb oldali konstans bármeddig növekedhet, az optimális megoldás nem változik (hiszen a lehetséges megoldások halmaza és a célfüggvény sem változik). A lehetséges csökkenés mértéke a harmadik feltétel $s_3$ *eltérésváltozójának* értéke az optimális megoldásban: $-x_1 + x_2 + s_3 = 4$, ahol $s_3 \geq 0$.
<br> Optimális megoldásban $x_1 = 6, x_2=6$, azaz $s_3 = 4$.
<br> Összességében tehát a jobb oldali konstans legfeljebb 4-gyel csökkenhet, és tetszőlegesen nőhet ahhoz, hogy az optimális megoldás optimális maradjon. Az optimális célfüggvény érték ekkor nem változik.

*Megjegyzés: ezáltal (korlátozottan) változó jobb oldali konstans esetén az LP újboli megoldása nélkül is meg tudjuk határozni az új optimális megoldást és hozzá tartozó célfüggvény értéket*

A definiált LpProblem object constraints.items() methodja segítségével az eltérésváltozók értékeit is le tudjuk olvasni.

In [ ]:
for name, c in varrono.constraints.items():
    print('s' + name + ' =', c.slack)

g) Tegyük fel, hogy a varrónőnek lehetősége van nadrágot is készíteni, melynek darabja 2 méter A, 5 méter B és 1 méter maradékanyagot igényel. Legalább milyen áron kell tudnia értékesíteni a nadrágot, hogy megérje azt is varrni?

*Megoldás:*
<br>az árnyékárak alapján a nadrág értékesítési ára legalább $2*6.666667+5*3.333333+1*0=30$, azaz legalább 30 ezer Ft kell legyen.

In [ ]:
#pip install pulp # ha még egyszer sem volt telepítve

from pulp import *

varrono = LpProblem('Varrónő', LpMaximize)

x1 = LpVariable('ing', lowBound=0 , cat = 'Continuous')
x2 = LpVariable('nyakkendo', lowBound=0 , cat = 'Continuous')
x3 = LpVariable('nadrag', lowBound=0 , cat = 'Continuous')

varrono += 20*x1 + 30*x2 + 25*x3

varrono += x1 + 3*x2 + 2*x3 <= 24, "A anyag"
varrono += 4*x1 + 3*x2 + 5*x3 <= 42, "B anyag"
varrono += - x1 + x2 + x3 <= 4, "maradék anyag"

print(varrono)

megoldas = varrono.solve()

if LpStatus[megoldas] == 'Optimal':
    print('Optimális megoldást találtunk.')
    print('Célfüggvény érték: z* = ', value(varrono.objective))
    print('Megoldás: ing* = ', value(x1), ', nyakkendo* = ', value(x2), ', nadrag* = ', value(x3))
elif megoldas == -1:
    print('Az LP-nek nincs megengedett megoldása.')
elif LpStatus[megoldas] == 'Unbounded':
    print('A célfüggvény nem korlátos.')
else:
    print('Az LP megoldása valamely problémába ütközött.')

Varrónő:
MAXIMIZE
20*ing + 25*nadrag + 30*nyakkendo + 0
SUBJECT TO
A_anyag: ing + 2 nadrag + 3 nyakkendo <= 24

B_anyag: 4 ing + 5 nadrag + 3 nyakkendo <= 42

maradék_anyag: - ing + nadrag + nyakkendo <= 4

VARIABLES
ing Continuous
nadrag Continuous
nyakkendo Continuous

Optimális megoldást találtunk.
Célfüggvény érték: z* =  300.0
Megoldás: ing* =  6.0 , nyakkendo* =  6.0 , nadrag* =  0.0


In [ ]:
print('Redukált költségek:')
for v in varrono.variables():
    print(v.name)
    print(v.dj)

Redukált költségek:
ing
-0.0
nadrag
-5.0
nyakkendo
-0.0


h) Milyen tartományban változhatnak az egyes célfüggvény-együtthatók értékei ahhoz, hogy az optimális megoldás ne változzon? Hogyan változik eközben az optimális célfüggvény érték?

*Megoldás:*
<br>A két korlát, melynek csúcspontjában az optimális megoldás található a $x_1 + 3x_2 \leq 24$ illetve a $4x_1 + 3x_2 \leq 42$. Ezek meredekségei rendre -1/3 és -4/3. A $20x_1 + 30x_2$ célfüggvény meredeksége -2/3.

Legyen most az ingek értékesítési ára egy $c_1$ ismeretlen paraméter. Egészen addig, amíg a célfüggvény meredeksége -1/3 és -4/3 között marad, addig az eredeti optimális megoldás továbbra is az marad. Tehát egész addig, amíg $c_1$ teljesíti, hogy

$$
-\frac{1}{3} \geq -\frac{c_1}{3} \geq -\frac{4}{3}
$$

Ez akkor teljesül, ha $1 \leq c_1 \leq 4$. Ezen tartományon belül az $x_1=6, x_2=6$ továbbra is optimális megoldás, az optimális célfüggvény értéke pedig $300 + (c_1-20)*6$.

Ugyanez a nyakkendők tucatjainak $c_2$ értékesítési ára esetén

$$
-\frac{1}{3} \geq -\frac{2}{c_2} \geq -\frac{4}{3}
$$

Azaz $1.5 \leq c_2 \leq 6$ tartományban a (6; 6) továbbra is optimális megoldás, az optimális célfüggvény érték $300 + (c_2-30)*6$.

*Megjegyzés: ezáltal (korlátozottan) változó célfüggvény együtthatók esetén az LP újboli megoldása nélkül is meg tudjuk határozni az új optimális megoldást és hozzá tartozó célfüggvény értéket*

### Gyakorló feladat: dualitás, árnyékárak, célfüggvény együtthatók

*2. feladat*

Térjünk vissza az 1. gyakorlat 1. feladatára, melyben katonákat képeztünk ki és páncélokat gyártottunk, és menjünk végig a fentebbi b)-h) feladatokon.

$$\color{blue}{
\begin{array}{rccc}
\max & \phantom{3}x_1 + 3x_2 & & \\
\text{f.h.} & \phantom{3}x_1 + 2x_2 & \leq & 12 \\
& 3x_1 + 2x_2 & \leq & 24 \\
& \phantom{3}x_1 - \phantom{2}x_2 & \geq & 0 \\
& \phantom{1}x_1, \phantom{1}x_2 & \geq & 0
\end{array}
}$$

A c) pontban ügyeljünk a korlátok összegzésénél a relációkra!
<br> A h) feladatpontot a következő módosítással oldjuk meg: Tegyük fel, hogy lehetőségünk van harckocsit is gyártani, melynek darabja 6 órát és 4 dollárt igényel. Legalább mekkora hadereje kell legyen egy harckocsinak, hogy megérje gyártanunk? A meghatározott érték bármennyi harckocsi legyártására érvényes?

b) Adjunk meg egy **felső korlátot** a célfüggvény értékére az *első korlát alapján*!

*Megoldás:*
<br>Az $x_1$ együtthatója a célfüggvényben és a korlátban is 1, viszont az $x_2$-é a célfüggvényben 3, míg a korlátban 2. Szorozzuk meg a korlátot 3/2-vel annak érdekében, hogy szét tudjuk bontani az első korlátot magára a célfüggvényre + egy nem-negatív érték:

$$
3/2x_1 + 3x_2 = \color{blue}{x_1 + 3x_2} + 0.5x_1 \leq 18
$$

Az optimális megoldásban felvett célfüggvény érték az első korlát miatt nem lehet nagyobb, mint 18.

c) Adjunk meg jobb felső korlátot a második és harmadik korlát használatával!

*Megoldás:*
<br>Az ellentétes relációk miatt nem tudjuk a két korlátot csakúgy összegezni: ahhoz, hogy felső korlátot kapjunk (és egyben az összegzéshez), a második korlát $\leq$ relációja alkalmas, de a második $\geq$ relációját meg kéne fordítanunk egy negatív számmal való szorzással.

Szorozzuk meg a második korlátot valamilyen $y_2$ nem-negatív értékkel, míg a harmadikat valamilyen $-y_3$ nem-pozitív számmal, majd összegezzük az így kapott korlátokat. Az eredmény:

$$
(3y_2 - y_3)x_1 + (2y_2 + y_3)x_2 \leq 24y_2
$$

Ahhoz, hogy szétbonthassuk a korábbi módon, $x_1$ együtthatójának (ami $3y_2 - y_3$) legalább 1-nek ($x_1$ célfgv. együtthatója), $x_2$-nél pedig a $2y_2 + y_3$ együtthatónak kell legalább 3-nak lennie. Mindezt úgy, hogy $y_2, y_3 \geq 0$ kell az összegzéshez.

Például $y_2 = 7/6, y_3 = 2/3$ esetén

$$17/6x_1 + 3x_2 = \color{blue}{x_1 + 3x_2} + 11/6x_1 \leq 16$$

d) Írjuk fel a legjobb korlátot meghatározó duális LP-t! Adjuk meg egy megoldását!

*Megoldás:* Innen adódik a **duális** feladat, mely egy LP segítségével megtalálja a legjobb korlátot biztosító $y$ változókat.

$$\color{blue}{
\begin{array}{rcccx}
\max & \phantom{3}x_1 + 3x_2 & & & \hspace{10mm} \color{red}{\text{(duál változók)}}\\
\text{f.h.} & \phantom{3}x_1 + 2x_2 & \leq & 12 & \hspace{10mm} \color{red}{(y_1)}\\
& 3x_1 + 2x_2 & \leq & 24 & \hspace{10mm} \color{red}{(y_2)}\\
& \phantom{3}x_1 - \phantom{2}x_2 & \geq & 0 & \hspace{10mm} \color{red}{(-y_3)}\\
& \phantom{1}x_1, \phantom{1}x_2 & \geq & 0
\end{array}
}$$

$$\color{red}{
\begin{array}{rcccx}
\min & 12y_1 + 24y_2 \phantom{+0y_3} & & & \hspace{10mm} \color{blue}{\text{(primál változók)}}\\
\text{f.h.} & \phantom{12}y_1 + \phantom{2}3y_2 - \phantom{0}y_3 & \geq & 1 & \hspace{10mm} \color{blue}{(x_1)}\\
& \phantom{1}2y_1 + \phantom{2}2y_2 + \phantom{0}y_3 & \geq & 3 & \hspace{10mm} \color{blue}{(x_2)}\\
& \phantom{24}y_1, \phantom{12}y_2, \phantom{0}y_3 & \geq & 0
\end{array}
}$$

* Primál vátozók száma = duál korlátok száma és fordítva
* primál célfüggvény együtthatók = duál jobb oldali konstansok és fordítva
* a primál együtthatómátrix *majdnem* a duál együtthatómátrix transzponáltja <br>(ha $\leq$-re rendeznénk a harmadik primál korlátot)

In [1]:
#pip install pulp # ha még egyszer sem volt telepítve

from pulp import *

model = LpProblem('Hadsereg', LpMaximize)

x1 = LpVariable('katona', lowBound=0 , cat = 'Continuous')
x2 = LpVariable('páncél', lowBound=0 , cat = 'Continuous')

model += x1 + 3*x2

model += x1 + 2*x2 <= 12, "idő"
model += 3*x1 + 2*x2 <= 24, "pénz"
model += x1 - x2 >= 0, "üres páncél"

print(model)

megoldas = model.solve()

if LpStatus[megoldas] == 'Optimal':
    print('Optimális megoldást találtunk.')
    print('Célfüggvény érték: z* = ', value(model.objective))
    print('Megoldás: katona* = ', value(x1), ', pancel* = ', value(x2))
elif megoldas == -1:
    print('Az LP-nek nincs megengedett megoldása.')
elif LpStatus[megoldas] == 'Unbounded':
    print('A célfüggvény nem korlátos.')
else:
    print('Az LP megoldása valamely problémába ütközött.')

    
i=0
for name, c in model.constraints.items():
    if (i == 2):
        print('y' + str(i+1) + ' =', -c.pi)
    else:
        print('y' + str(i+1) + ' =', c.pi)
    i += 1

Hadsereg:
MAXIMIZE
1*katona + 3*páncél + 0
SUBJECT TO
idő: katona + 2 páncél <= 12

pénz: 3 katona + 2 páncél <= 24

üres_páncél: katona - páncél >= 0

VARIABLES
katona Continuous
páncél Continuous

Optimális megoldást találtunk.
Célfüggvény érték: z* =  16.0
Megoldás: katona* =  4.0 , pancel* =  4.0
y1 = 1.3333333
y2 = -0.0
y3 = 0.33333333


e) Értelmezzük az egyes árnyékárakat! A rendelkezésünkre álló erőforrások közül melyik mennyire értékes? Határozzuk meg melyek az aktív és inaktív feltételek!

*Megoldás:*
<br>Egy plusz óra $y_1 = 4/3$-dal növelné a haderőnket, míg ugyanez plusz 1 dollár esetében $y_2 = 0$. A harmadik árnyékár nehezebben értelmezhető természetes módon: ha egy selejtes páncélt gyártanánk, az $y_3 = 0.33333333$-mal csökkentené a haderőnket.
<br> Az első és harmadik feltétel aktív, a második feltétel inaktív.

f) Az inaktív feltételek esetén: a jobb oldali konstans milyen intervallumban változhat ahhoz, hogy az optimális megoldás ne változzon? Hogyan változik eközben az optimális célfüggvény érték?

*Megoldás:*
<br>$\leq$ korlát lévén a jobb oldali konstans bármeddig növekedhet, az optimális megoldás nem változik (hiszen a lehetséges megoldások halmaza és a célfüggvény sem változik). A lehetséges csökkenés mértéke a második feltétel $s_2$ *eltérésváltozójának* értéke az optimális megoldásban: $3x_1 + 2x_2 + s_2 = 24$, ahol $s_2 \geq 0$.
<br> Optimális megoldásban $x_1 = 4, x_2=4$, azaz $s_2 = 4$.
<br> Összességében tehát a jobb oldali konstans legfeljebb 4-gyel csökkenhet, és tetszőlegesen nőhet ahhoz, hogy az optimális megoldás optimális maradjon. Az optimális célfüggvény érték ekkor nem változik.

In [3]:
i=0
for name, c in model.constraints.items():
    print('s' + str(i+1) + ' =', c.slack)
    i += 1

s1 = -0.0
s2 = 4.0
s3 = 1e-12


g) Milyen tartományban változhatnak az egyes célfüggvény-együtthatók értékei ahhoz, hogy az optimális megoldás ne változzon? Hogyan változik eközben az optimális célfüggvény érték?

*Megoldás:*
<br>A két korlát, melynek csúcspontjában az optimális megoldás található a $x_1 + 2x_2 \leq 12$ illetve az $x_1 - x_2 \geq 0$. Ezek meredekségei rendre -1/2 és 1. Az $x_1 + 3x_2$ célfüggvény meredeksége -1/3.

Legyen most a katona hadereje egy $c_1$ ismeretlen paraméter. Egészen addig, amíg a célfüggvény meredeksége -1/2 és 1 között marad, addig az eredeti optimális megoldás továbbra is az marad. Tehát egész addig, amíg $c_1$ teljesíti, hogy

$$
-\frac{1}{2} \leq -\frac{c_1}{3} \leq 1
$$

Ez akkor teljesül, ha $1.5 \geq c_1 \geq -3$. Ezen tartományon belül az $x_1=4, x_2=4$ továbbra is optimális megoldás, az optimális célfüggvény értéke pedig $16 + (c_1-1)*4$.

Ugyanez a nyakkendők tucatjainak $c_2$ értékesítési ára esetén

$$
-\frac{1}{2} \leq -\frac{1}{c_2} \leq 1
$$

Azaz $-1 \leq c_2 \leq 2$ tartományban a (4; 4) továbbra is optimális megoldás, az optimális célfüggvény érték $16 + (c_2-3)*4$.

h) Tegyük fel, hogy lehetőségünk van harckocsit is gyártani, melynek darabja 6 órát és 4 dollárt igényel. Legalább mekkora hadereje kell legyen egy harckocsinak, hogy megérje gyártanunk? A meghatározott érték bármennyi harckocsi legyártására érvényes?

*Megoldás:*
<br>az árnyékárak alapján a harckocsi hadereje legalább $6*1.333333+4*0=8$, azaz legalább 8-as hadereje kell, hogy legyen. A második korlát eltérésváltozójának értéke éppen 4, azaz 1 harckocsi legyártását követően a második korlát 0-s árnyékára már nem feltétlen marad érvényben. Tehát a válasz nem, legfeljebb 1 harckocsi legyártásáról tudjuk ezt megállapítani.

## Mátrixjátékok és dualitás

Térjünk vissza a mátrixjátékokra: nem véletlen, hogy a játék értékét meg tudjuk határozni egyaránt a Sor és az Oszlop feladatából is. A két feladat majdnem teljesen egymás **duálisa**.

$$\color{blue}{
\begin{array}{cccc}
\hspace{20mm} \max v & & & \hspace{10mm} \color{red}{\text{(duál változók)}}\\
\phantom{-0x_1} - \phantom{1}x_2 + 2x_3 - v & \geq & 0 & \hspace{10mm} \color{red}{(-y_1)}\\
-1x_1 + 2x_2 - 3x_3 - v & \geq & 0 & \hspace{10mm} \color{red}{(-y_2)}\\
\phantom{+}2x_1 - 3x_2 + 4x_3 - v & \geq & 0 & \hspace{10mm} \color{red}{(-y_3)}\\
-3x_1 + 4x_2 - 5x_3 - v & \geq & 0 & \hspace{10mm} \color{red}{(-y_4)}\\
\phantom{1}x_1 + \phantom{1}x_2 + \phantom{1}x_3 \phantom{+0v} & = & 1 & \hspace{10mm} \color{red}{(w)}\\
\phantom{1}x_1, \phantom{1}x_2, \phantom{1}x_3, \phantom{,0v} & \geq & 0
\end{array}
}$$

Az együtthatómátrix
$$
\color{blue}{
\left[ \begin{array}{ccc|c}
 \phantom{-}0 & -1 & \phantom{-}2 & -1 \\
 -1 & \phantom{-}2 & -3 & -1\\
 \phantom{-}2 & -3 & \phantom{-}4 & -1 \\
 -3 & \phantom{-}4 & -5 & -1 \\ \hline
 \phantom{-}1 & \phantom{-}1 & \phantom{-}1 & \phantom{-}0
\end{array} \right]
}$$
transzponáltja lesz a duál feladat együtthatómátrixa.
$$
\color{red}{
\left[ \begin{array}{cccc|c}
 \phantom{-}0 & -1 & \phantom{-}2 & -3 & 1\\
 -1 & \phantom{-}2 & -3 & \phantom{-}4 & 1\\
 \phantom{-}2 & -3 & \phantom{-}4 & -5 & 1\\ \hline
 -1 & -1 & -1 & -1 \phantom{-} & 0
\end{array} \right]
}$$
Az eredeti feladat célfüggvény együtthatói $[0 \ 0 \ 0 \ 1]$ és jobb oldali konstansai $[0 \ 0 \ 0 \ 0 \ 1]$ lesznek a duál feladat jobb oldali konstansai és célfüggvényegyütthatói. A maximalizálás minimalizálásra vált: $\color{red}{\min w}$

Az $\color{blue}{x_1, x_2, x_3}$ változók előjelkikötései a duál feladat első három korlátozó feltételeiben egyenlőtlenséget, $\color{blue}v$ kötetlensége az utolsó feltételben egyenlőséget eredményez.

$$\color{red}{
\begin{array}{ccccc}
\hspace{20mm} \min w & & & \hspace{10mm} \color{blue}{primál\ változók}\\
\phantom{-0y_1+1}y_2 - 2y_3 + 3y_4 + w & \geq & 0 & \hspace{10mm} \color{blue}{(x_1)}\\
\phantom{-1}y_1 - 2y_2 + 3y_3 - 4y_4 + w & \geq & 0 & \hspace{10mm} \color{blue}{(x_2)}\\
-2y_1 + 3y_2 - 4y_3 + 5y_4 + w & \geq & 0 & \hspace{10mm} \color{blue}{(x_3)}\\
\phantom{1}-y_1 - \phantom{1}y_2 - \phantom{1}y_3 - \phantom{1}y_4 \phantom{+0w} & = & 1 & \hspace{10mm} \color{blue}{(v)}\\
\phantom{1}-y_1, \phantom{1}-y_2, \phantom{1}-y_3, \phantom{1}-y_4, \phantom{,0w} & \geq & 0
\end{array}
}$$

Innen, a könnyebb értelmezhetőség végett $-y_i$-ből $y_i$-re történő változótranszformációval megkapjuk Oszlop feladatát az eredeti formájában.

In [ ]:
from pulp import *

sor = LpProblem('sor', LpMaximize)

x = LpVariable.dicts('x', range(3), lowBound=0, cat='Continuous')
v = LpVariable('v', cat='Continuous')

A = [[0, -1, 2, -3],[-1, 2, -3, 4],[2, -3, 4, -5]]

sor += v

for i in range(len(A[0])):
    sor += pulp.lpSum([A[j][i]*x[j] for j in range(len(A))]) - v >= 0, "oszlop stratégia " + str(i+1)
    
sor += pulp.lpSum([x[i] for i in range(3)]) == 1, "valószínűség változók"

print(sor)

megoldas = sor.solve()

if LpStatus[megoldas] == 'Optimal':
    print('Optimális megoldást találtunk.')
    print('Játék értéke: v* =', value(sor.objective))
    print('Sor kevert stratégiája:')
    for i in range(3):
        print('x' + str(i+1) + ' =', value(x[i]))
    print('Oszlop kevert stratégiája:')
    i=0
    for name, c in sor.constraints.items():
        print('y' + str(i+1) + ' =', -c.pi)
        i += 1
        if (i == len(A[0])):
            break
elif megoldas == -1:
    print('Az LP-nek nincs megengedett megoldása.')
elif LpStatus[megoldas] == 'Unbounded':
    print('A célfüggvény nem korlátos.')
else:
    print('Az LP megoldása valamely problémába ütközött.')